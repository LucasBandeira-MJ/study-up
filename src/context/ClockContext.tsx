import { createContext, useEffect, useState } from "react";

let clockTimeout: NodeJS.Timeout

interface ClockContextData {
    time: number
    hours: number
    minutes: number
    seconds: number
    isActive: boolean
    startClock: () => void
    pauseClock: () => void
    stopClock: () => void
}

interface ClockProviderProps {
    children: React.ReactNode
}

export const ClockContext = createContext({} as ClockContextData)

export const ClockProvider = ({children}: ClockProviderProps) => {
    const [time, setTime] = useState<number>(0)
    const [isActive, setIsActive] = useState<boolean>(false)

    useEffect(() => {
        if(isActive) {
            clockTimeout = setTimeout(() => {
                setTime(time => time + 1)
            }, 1000)
        }
    }, [time, isActive])

    const startClock = () => {
        setIsActive(true)
    }

    const pauseClock = () => {
        clearInterval(clockTimeout)
        setIsActive(false)
    }

    const stopClock = () => {
        clearInterval(clockTimeout)
        setIsActive(false)
    }


    const hours = Math.floor(time / 3600)
    const minutes = Math.floor(time / 60 - hours * 60)
    const seconds = Math.floor(time % 60)

    return (
        <ClockContext.Provider value={{
            time,
            hours, 
            minutes, 
            seconds,
            isActive,
            startClock,
            pauseClock,
            stopClock
        }}>
            {children}
        </ClockContext.Provider>
    )
}