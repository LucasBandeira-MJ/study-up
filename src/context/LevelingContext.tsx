import { createContext, ReactNode, useEffect, useState } from "react";

let clockTimeout: NodeJS.Timeout

interface LevelingContextData {
    time: number
    isClockActive: boolean
    showLevelingModal: boolean
    level: number
    currentExperience: number
    experienceToNextLevel: number
    totalExperience: number
    totalLevelUps: number
    handleClock: () => void
    stopClock: () => void
    closeLevelingModal: () => void
}

interface LevelingProviderProps {
    children: ReactNode
}

export const LevelingContext = createContext({} as LevelingContextData)

export const LevelingProvider = ({ children }: LevelingProviderProps) => {
    const [time, setTime] = useState<number>(0)
    const [isClockActive, setIsClockActive] = useState<boolean>(false)
    const [showLevelingModal, setShowLevelingModal] = useState<boolean>(false)
    const [level, setLevel] = useState<number>(1)
    const [currentExperience, setCurrentExperience] = useState<number>(0)
    const [totalExperience, setTotalExperience] = useState<number>(0)
    const [totalLevelUps, setTotalLevelUps] = useState<number>(0)

    const experienceToNextLevel = Math.pow((level + 1) * 2, 2)

    const handleClock = () => {
        if(isClockActive) {
            clearInterval(clockTimeout)
            setIsClockActive(false)
        } else {
            setIsClockActive(true)
        }
    }

    const stopClock = () => {
        clearInterval(clockTimeout)
        setIsClockActive(false)
        setShowLevelingModal(true)
    }

    const closeLevelingModal = () => {
        setShowLevelingModal(false)
        setTime(0)
    }


    const levelUp = () => {
        setLevel(level => level + 1)
        setTotalLevelUps(totalLevelUps => totalLevelUps + 1)
    }


    useEffect(() => {
        if(isClockActive) {
            clockTimeout = setTimeout(() => {
                setTime(time => time + 1)
            }, 1000)
        }
    }, [time, isClockActive])

    useEffect(() => {
        let finalExperience = !!time ?  currentExperience + 1 : currentExperience
        let totalExp = !!time ? totalExperience + 1 : totalExperience

        if(finalExperience >= experienceToNextLevel) {
            finalExperience = finalExperience - experienceToNextLevel
            levelUp()
        }

        setCurrentExperience(finalExperience)
        setTotalExperience(totalExp)
    }, [time])


    return (
        <LevelingContext.Provider value={{
            time,
            isClockActive,
            showLevelingModal,
            level,
            currentExperience,
            experienceToNextLevel,
            totalExperience,
            totalLevelUps,
            handleClock,
            stopClock,
            closeLevelingModal
        }}>
            {children}
        </LevelingContext.Provider>
    )
}

