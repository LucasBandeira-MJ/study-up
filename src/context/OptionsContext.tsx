import { createContext, useState } from "react";
import { OptionsModal } from "../components/OptionsModal";

interface OptionsProviderProps {
    children: React.ReactNode
}

interface OptionsContextData {
    isDarkMode?: boolean
    isOptionsModalOpen: boolean
    toggleTheme: () =>  void
    openOptionsModal: () => void
    closeOptionsModal: () => void
}

export const OptionsContext = createContext({} as OptionsContextData)

export const OptionsProvider = ({children}: OptionsProviderProps) => {
    const [isOptionsModalOpen, setIsOptionsModalOpen] = useState(false)
    const [isDarkMode, setIsDarkMode] = useState(false)

    const toggleTheme = () => {
        setIsDarkMode(theme => !theme)
    }

    const openOptionsModal = () => setIsOptionsModalOpen(true) 
    const closeOptionsModal = () => setIsOptionsModalOpen(false) 

    return (
        <OptionsContext.Provider value={{ 
            isDarkMode, 
            isOptionsModalOpen,
            toggleTheme,
            openOptionsModal,
            closeOptionsModal
        }}>
            {children}
            <OptionsModal />
        </OptionsContext.Provider>   
    )
}