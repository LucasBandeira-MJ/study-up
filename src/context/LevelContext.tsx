import { createContext, useContext, useEffect, useState } from "react";
import { LevelingModal } from "../components/LevelingModal";
import { ClockContext } from "./ClockContext";

interface LevelContextData {
    level: number
    currentExperience: number
    experienceToNextLevel: number
    totalExperience: number
    totalLevelUps: number
}

interface LevelProviderProps {
    children: React.ReactNode
}

export const LevelContext = createContext({} as LevelContextData)

export const LevelProvider = ({children}: LevelProviderProps) => {
    const { time } = useContext(ClockContext)

    const [level, setLevel] = useState<number>(1)
    const [currentExperience, setCurrentExperience] = useState<number>(0)
    const [totalExperience, setTotalExperience] = useState<number>(0)
    const [totalLevelUps, setTotalLevelUps] = useState<number>(0)

    const experienceToNextLevel = Math.pow((level + 1) * 2, 2)

    const levelUp = () => {
        setLevel(level => level + 1)
        setTotalLevelUps(totalLevelUps => totalLevelUps + 1)
    }

    useEffect(() => {
        console.log('uh')
        let finalExperience = !!time ?  currentExperience + 1 : currentExperience
        let totalExp = !! time ? totalExperience + 1 : totalExperience

        if(finalExperience >= experienceToNextLevel) {
            finalExperience = finalExperience - experienceToNextLevel
            levelUp()
        }

        setCurrentExperience(finalExperience)
        setTotalExperience(totalExp)
    }, [time])


    return (
        <LevelContext.Provider value={{
            level,
            currentExperience,
            experienceToNextLevel,
            totalExperience,
            totalLevelUps
        }}>
            {children}
            <LevelingModal />
        </LevelContext.Provider>
    )
}