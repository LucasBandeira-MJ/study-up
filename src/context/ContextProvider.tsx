import { LevelingProvider } from "./LevelingContext"
import { OptionsProvider } from "./OptionsContext"
import { SessionProvider } from 'next-auth/react'

interface ContextProviderProps {
    children: React.ReactNode
}

export const ContextProvider = ({children}: ContextProviderProps) => {
    return (
        <SessionProvider>
            <OptionsProvider>
                <LevelingProvider>
                    {children}
                </LevelingProvider>
            </OptionsProvider>
        </SessionProvider>
    )
}