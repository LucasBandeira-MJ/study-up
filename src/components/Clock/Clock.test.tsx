import { render, screen } from "@testing-library/react"
import { Clock } from "."
import renderer from 'react-test-renderer'
import { LevelingContext } from "../../context/LevelingContext"

describe("Clock component", () => {
    it('should render correctly', () => {
        render(
            <LevelingContext.Provider value={{ hours: 2, minutes: 53, seconds: 14 } as any}>
                <Clock />
            </LevelingContext.Provider>
        )

        const hoursLeft = screen.getByText('0')
        const hoursRight = screen.getByText('2')
        const minutesLeft = screen.getByText('5')
        const minutesRight = screen.getByText('3')
        const secondsLeft = screen.getByText('1')
        const secondsRight = screen.getByText('4')

        expect(hoursLeft).toBeInTheDocument()
        expect(hoursRight).toBeInTheDocument()
        expect(minutesLeft).toBeInTheDocument()
        expect(minutesRight).toBeInTheDocument()
        expect(secondsLeft).toBeInTheDocument()
        expect(secondsRight).toBeInTheDocument()
    })

    it('should match the snapshot', () => {
        const snapshot = renderer
            .create(
                <LevelingContext.Provider value={{hours: 5, minutes: 23, seconds: 45} as any}>
                    <Clock />
                </LevelingContext.Provider>
            ).toJSON()

            expect(snapshot).toMatchSnapshot()
      });

})