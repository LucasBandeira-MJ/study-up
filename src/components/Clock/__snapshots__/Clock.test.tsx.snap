// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock component should match the snapshot 1`] = `
Array [
  <div
    className="container"
  >
    <div>
      <span>
        0
      </span>
      <span>
        5
      </span>
    </div>
    <span>
      :
    </span>
    <div>
      <span>
        2
      </span>
      <span>
        3
      </span>
    </div>
    <span>
      :
    </span>
    <div>
      <span>
        4
      </span>
      <span>
        5
      </span>
    </div>
  </div>,
  <div
    className="buttonContainer"
  >
    <button
      className="
            clockButton 
            undefined
        "
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Start
       Timer
    </button>
    <button
      className="
            clockButton 
            stopButton
        "
      onClick={[Function]}
      type="button"
    >
      Stop
       Timer
    </button>
  </div>,
]
`;
